srai/neighbourhood/neighbourhood.py

```diff
@@ -0,0 +1,93 @@
+"""Neighbourhood interface."""
+
+from abc import ABC, abstractmethod
+from queue import Queue
+from typing import Generic, Set, Tuple, TypeVar
+
+from functional import seq
+
+IndexType = TypeVar("IndexType")
+
+
+class Neighbourhood(ABC, Generic[IndexType]):
+    """Neighbourhood interface."""
+
+    @abstractmethod
+    def get_neighbours(self, index: IndexType) -> Set[IndexType]:
+        """
+        Get the direct neighbours of a region using its index.
+
+        Args:
+            index (Any): Unique identifier of the region.
+                Dependant on the implementation.
+
+        Returns:
+            Set[Any]: Indexes of the neighbours.
+        """
+        pass
+
+    def get_neighbours_up_to_distance(self, index: IndexType, distance: int) -> Set[IndexType]:
+        """
+        Get the neighbours of a region up to a certain distance.
+
+        Args:
+            index (Any): Unique identifier of the region.
+                Dependant on the implementation.
+            distance (int): Maximum distance to the neighbours.
+
+        Returns:
+            List[Any]: Indexes of the neighbours.
+        """
+        neighbours_with_distances = self._get_neighbours_with_distances(index, distance)
+        neighbours: Set[IndexType] = seq(neighbours_with_distances).map(lambda x: x[0]).to_set()
+        if index in neighbours:
+            neighbours.remove(index)
+        return neighbours
+
+    def get_neighbours_at_distance(self, index: IndexType, distance: int) -> Set[IndexType]:
+        """
+        Get the neighbours of a region at a certain distance.
+
+        Args:
+            index (Any): Unique identifier of the region.
+                Dependant on the implementation.
+            distance (int): Distance to the neighbours.
+
+        Returns:
+            List[Any]: Indexes of the neighbours.
+        """
+        neighbours_up_to_distance = self._get_neighbours_with_distances(index, distance)
+        neighbours_at_distance: Set[IndexType] = (
+            seq(neighbours_up_to_distance)
+            .filter(lambda x: x[1] == distance)
+            .map(lambda x: x[0])
+            .to_set()
+        )
+        if index in neighbours_at_distance:
+            neighbours_at_distance.remove(index)
+        return neighbours_at_distance
+
+    def _get_neighbours_with_distances(
+        self, index: IndexType, distance: int
+    ) -> Set[Tuple[IndexType, int]]:
+        visited = set()
+        visited_with_distances = set()
+        to_visit: Queue[Tuple[IndexType, int]] = Queue()
+
+        visited.add(index)
```

Comment:
Why add this first region twice? Here and in `while` loop?

Reply:
I changed it a couple of times and ended up with this. I simplified it, so I hope it now looks better.

Reply:
Thanks!